### ISSUE TITLE
Chained conditionals: how they work, how they *should* work and how we can optimize them

### ISSUE THREAD

**Issue by drkameleon** [arturo-lang/arturo#1247]
This issue is not properly speaking a bug, but its most definite goal is to deal with - perhaps by design - features which often lead to issues: namely chained conditionals, or more concretely `if?`/`else` structures & `case`/`when?`/.../`else` structures.

Also, we've thoroughly discussed the issue with @RickBarretto, but I think it'd be better to have all of this in the open, not only as a means of exchanging ideas publicly, but also so that we can gather all the necessary information/tests/etc in one place. 🚀

**Comment by drkameleon** [comment]
## The case of `if?`-`else`

First and foremost, a critical detail: in most programming languages, this would be considered a statement, while in Arturo it's not *one* statement; we actually have *two*: the first one being the `if?` (with its two arguments: the condition and the block), followed by `else` (with its single argument: a block).

Technically, this is like writing:

```red
if? x = 2 [print "true"] ; first command
else [print "false"].    ; second command
```

We might tend to see them as one structure, but we have to be very careful not to. They are 2 commands ❗ 

### How *was* that working?

The `if?` command - [the way its implemented in our library](https://github.com/arturo-lang/arturo/blob/master/src/library/Core.nim#L516-L520), at least - is supposed to act like a normal `if`, only instead of just checking the veracity of the given condition and executing the block (or not), it also returns the value of that condition (true/false). And in that sense, it abides by the unwritten rule in Arturo that all predicates (= functions that return a Logical, true/false, value) end with a question mark. So far, so good.

Now, how does this `else` work? Since it always follows an `if?` statement (or at least that's what's expected), it actually pops the last Logical value from the stack (the one that has been pushed by the preceding `if?`) and - based on that - decides whether it should procede and execute its own block, or not.

[Its implementation](https://github.com/arturo-lang/arturo/blob/master/src/library/Core.nim#L424-L426) is also quite easy to grasp.

### How is it working *now*? (or at least most of the time)

As much as I kept taking all of the above for granted, to my own suprise, someone decided to go on and implement an AST tree, that stands between our parsed values and the generated bytecode. And not only that: he also decided to optimize this `if?`-`else` structure away! 🤣 

So, when I tried to see the produced bytecode for a very simple `if?`-`else` block, the bytecode I saw had no calls to `if?` or `else` at all. It was actually, a very clean, Assembly-style block of logical jumps.

**Example input:**

```red
ib: function [x]-> inspect to :bytecode x

ib [
    if? x=2 -> print x
    else -> print x*2
]
```

**Example output:**

```red
[ :bytecode
        ================================
         DATA
        ================================
        0: x :word

        ================================
         CODE
        ================================
        consti2
        load0
        jmpifne              @4
        load0
        print
        goto                 @4
        consti2
        load0
        mul
        print
        end
]
```

**So, what does this bytecode do?**

- we load up the constant `2` and the variable `x`
- we call `jmpifne` (= jump if not equal); what is not equal? the topmost stack values (in our case: `2` and `x`) and jump to where? jump ahead for 4 bytes (now, this is quite tricky to explain, since it depends on how many bytes each command takes up,  but in our case this jmp points to the second `consti2`
- if the values are NOT not-equal (= how to say "they are equal", and go totally crazy) the command is ignored and it proceed to `print x`
- then it encounters this `goto @4` which effectively makes it jump to 4 bytes ahead (= bypass the `else` block) and continue as usual
- now, if the values where actually not-equal, it would jump to `consti2` and `print 2*x` instead.

All of this mess is just a far, *far* more performant and closer-to-the-metal way of implementing an if/else condition.

**So, why write all this?**

Because apparently, some of these leave hanging values onto the stack. And from what I've written above (and the AST optimization), it most definitely appears the `if?`-`else` is not the culprit. (Or, at least, it *shouldn't* be, if the AST optimizations work at *all* times, without a single exception - which has to be proven in practice!)

**Comment by drkameleon** [comment]
As a useful and important sidenote to the above explanation:

Arturo already has an one-statement `if?`-`else` alternative; and that would be [`switch`](https://arturo-lang.io/documentation/library/core/switch/) (or `?`), which is very similar - if not identical - to [Rebol's `either`](http://www.rebol.com/docs/words/weither.html).

Now, why have this too? Do we just need another function simply to do the exact same thing?

It's not that simple. 

What it comes down to again is the fact that we would be talking about *one* single statement (instead of two). And that means that, using `switch`, we can do things like:

```red
x: 0

y: switch x=0 -> 1 -> 2 ; if x is 0, set y to 1, otherwise set it to 2
```

or using our `?` alias:

```red
x: 0
y: (x=0)? -> 1 -> 2
```

...which looks good and tidy; plus, if we were to use `if?`-`else` for that, I don't think I could come up with a different, safe solution, other than:

```red
y: 0
x: 0
if? x = 0 -> y: 1
else -> y: 2
```

(Verbose... too verbose! lol)

**Comment by drkameleon** [comment]
**Note no 2:**

```red
b: function [x]-> inspect to :bytecode x

identical?: function [z][
    one? unique map @z 'a -> to :bytecode a
]

ib [
    if? x=2 -> print x
    else -> print x*2
]

print identical? [
    [
        if? x=2 -> print x
        else -> print x*2
    ]
    [
        if? x=2 -> print x

        else -> print x*2
    ]
    [
        if? x=2 
            -> print x
        else 
            -> print x*2
    ]
    [
        if? 
        x=2 
        -> print x
        else 
        -> print x*2
    ]
    [
        if? x=2 -> print x
        
        ; some comment
        ; another comment
        
        else -> print x*2
    ]
    [
        switch x=2 -> print x
                   -> print x*2
    ]
    [
        (x=2)? -> print x
               -> print x*2
    ]
]
```

Result:

```red
[ :bytecode
        ================================
         DATA
        ================================
        0: x :word

        ================================
         CODE
        ================================
        consti2
        load0
        jmpifne              @4
        load0
        print
        goto                 @4
        consti2
        load0
        mul
        print
        end
]
true
```

(Not only do newlines not affect the produced bytecode, but... apparently `switch` is optimized in the exact same fashion!)

**Comment by programandala-net** [comment]
> Now, how does this `else` work? Since it always follows an `if?` statement (or at least that's what's expected), it actually pops the last Logical value from the stack (the one that has been pushed by the preceding if?) and - based on that - decides whether it should procede and execute its own block, or not.

At first I found a bit confusing the fact the documentation doesn't mention the relation between `if?` and `else` explicitly, but only implicitly in the code examples. Also, the [page about `else`](https://arturo-lang.io/documentation/library/core/else/)  does not include `if?` in the list of its related keywords... but `if`!

But at the end everything became clear. `else` is very versatile: it can be used by itself, provided you know what's on the stack, like in this Forth-like code I wrote to simulate a BASIC numerical `input`:

```
until [
	numberString: input "Enter a number: "
	dup numeric? numberString else [ print "Number expected. Retry." ]
] [ ]
number: floor to :floating numberString
```

**Comment by drkameleon** [comment]
> At first I found a bit confusing the fact the documentation doesn't mention the relation between `if?` and `else` explicitly, but only implicitly in the code examples. Also, the [page about `else`](https://arturo-lang.io/documentation/library/core/else/) does not include `if?` in the list of its related keywords... but `if`!

The truth is the documentation could be lacking in various cases. And that is one of them.

Btw, you're more than welcome to make a PR! 😉

**Comment by hostilefork** [comment]
Hello there, [long time no speak](https://forum.rebol.info/t/arturo-rebol-inspired-language/1370).

I hadn't checked in on your project for a while, but I did recently.  I am in particular impressed with [how many examples](https://arturo-lang.io/documentation/examples/) you have put together.  It's something other Rebolverse projects should aspire to, and you've certainly raised the bar for that!

So I decided to poke around a bit and learn more about the state of things with your work.

I will admit to not understanding completely what the language goals are (vs. being a personal project written in Nim, to explore the nuances of synthetic exercises in a layer above Nim, that might appeal to Nim programmers, (?))

But...

> Arturo already has a one-statement, if?-else alternative; and that would be [switch](https://arturo-lang.io/documentation/library/core/switch/) (or ?), which is very similar - if not identical - to [Rebol's either](http://www.rebol.com/docs/words/weither.html).

Calling this SWITCH seems pretty much like "difference for the sake of being different".

I'd kind of like you to understand more about the medium from which you are borrowing influences--in addition to not randomizing the terminology.  There's a lot to learn, e.g. why IF doesn't need a question mark after it...ever.  Could I possibly interest you in re-engaging and discussing the tech behind a Rebol-ish language?

**Comment by drkameleon** [comment]
@hostilefork 

Good to see you around!

Thanks a lot for your kind words. The project is probably the one I'm most proud of, that's why I keep at it and working actively on its development. Of course, its progress and evolution wouldn't have been the same if it weren't for the people that have contributed to it (and not just code) - so, I consider it a team effort. 🚀 

The "trick" of having a rather wide array of examples to offer has actually been an interesting one: apart from promoting the language, it has definitely helped in spotting early issues, missing features, etc; not to mention that it also provides *us* with material for the unit-tests. So, that would be a win-win-win.

As for the bigger idea, my initial idea was sth that would be "batteries-included" and easily package-able for the 3 major OS'es. It most certainly wasn't about Nim (the language was merely picked because it serves the goals of the project). And it's also not just a personal, hobby language anymore - as a matter of fact, it has been used in production for at least the past two years 😉  Perhaps an even better way of understanding the direction the project has taken - surely more efficient than me talking about aspirations - would be to have a look into our Package Manager, published projects in general, projects by other people, etc. Discord would also be a good place to see what is going on - since its our main place were you usually exchange ideas.

Re: Rebol.

As I've said again, I don't consider Arturo a... new Rebol, or a better Rebol, or anything like that. It has been a new idea since day one. Now, similarities do exist and that's why I may look up how function X may be called in Rebol... (but not much more, than I'd do it e.g. in OCaml, or Haskell, or Ruby - which as a matter of fact is the no.1 language that has probably influenced Arturo. You would be surprised how many languages I skim through for inspiration purposes, especially when it comes to naming things).

> Calling this SWITCH seems pretty much like "difference for the sake of being different".

Interesting. Obviously, it's a matter of perspective: You see this as an `either` that I decided to rename as `switch`... to differ. Well, for me it was never an `either`. I just looked it up to see if there was an if-else, in one command, and what the name was. Only to find an even more horrible name... lol (and no, I don't particularly love `switch` either). 

All that being said, I am obviously open to exchanging ideas. Does it mean that I'll accept them? Not really. But even ideas that you initially disagree with may have sth to offer. So,... totally welcome! 😉

**Comment by hostilefork** [comment]
> have a look into our [Packager](https://pkgr.art/)

Very nice-looking pages! (as pretty much all of yours are), though somewhat lacking in "discoverability".  One can't tell how many packages there are up front.  And once you click through to a package there are tags...but clicking on a tag doesn't get you a list of the packages with that tag.

Even just a way to get a flat list would be helpful.

> You see this as an either that I decided to rename as switch

Well... actually, I saw it as an either that you found out an established name for and didn't change it... instead giving it a name that has a far more established meaning for something else:

  https://en.wikipedia.org/wiki/Switch_statement

But you have put forth the explanation of not adopting EITHER because you don't like it:

> I just looked it up to see if there was an if-else, in one command, and what the name was. Only to find an even more horrible name

My opinion is that EITHER is unusually fitting, as it is commonly used in English *specifically* for binary choices.  From the Cambridge dictionary:

    EITHER - determiner, pronoun, conjunction
    used when referring to a choice between two possibilities

    "Would you like the metal or plastic one?" "Either will do."
    "Either candidate would be ideal for the job."
    "Do you prefer pork or beef?" "I don't like either."
    "You can get there by train or bus, either way it'll take an hour"
    "We can either eat now or after the show, it's up to you."
    "Either you leave now or I call the police!"

It's tough to find a nice word that carries that one-of-two choices connotation that works quite as well as EITHER does.

    either you.action = 'leave [close door] [call police]

> As I've said again, I don't consider Arturo a... new Rebol, or a better Rebol, or anything like that. It has been a new idea since day one.

No...Arturo is not Rebol, you don't have to convince me.  :-)  *However* the basics overlap enough to the point where I think interesting designs from Rebol could be brought to bear upon it.

It's a missed opportunity--I believe--for you to have done this much work on something *Rebol-like* and yet not get into the "fun" of those designs.  (If I seem puzzled about what you are doing, it is because I know how close you are to those things...but only scratching the surface.)

>Discord would also be a good place to see what is going on - since its our main place were we usually exchange ideas.

Not my typical medium, but I can look at it...

**Comment by drkameleon** [comment]
> Very nice-looking pages! (as pretty much all of yours are), though somewhat lacking in "discoverability". One can't tell how many packages there are up front. And once you click through to a package there are tags...but clicking on a tag doesn't get you a list of the packages with that tag.
> 
> Even just a way to get a flat list would be helpful.

Thanks a lot! 😄 

I've been meaning to start tweaking things but decided to put it off until we had more packages, so that we could... see what was missing. Have a look (I wrote this literally yesterday): https://discord.com/channels/765519132186640445/829324913097048065/1291021050364166147

> Well... actually, I saw it as an either that you found out an established name for and didn't change it... instead giving it a name that has a far more established meaning for something else:
> 
> https://en.wikipedia.org/wiki/Switch_statement

I know... As I said, I don't really like it myself. And yes, obviously even *I* think of a... C `switch` when seeing it.
However, on a positive note, it's possible that we end up avoiding it altogether (which could even be an argument for `either` lol): Arturo supports symbols; every symbol may/can be "aliased" to a function (prefix or infix style)... so in the end, I tend to use `?` like:

```
print (3 < 10)? -> "Yes!"
                -> "No :("
```

The same could be true for `coalesce` for example: it *is* descriptive although I wouldn't especially consider it the most great-sounding choice in the world. Now, do we actually use it? Not really. 

```
option: (attr 'opt) ?? "default value"
```

(and yes, that would still be our `coalesce`) 

> No...Arturo is not Rebol, you don't have to convince me. :-) However the basics overlap enough to the point where I think interesting designs from Rebol could be brought to bear upon it.
> 
> It's a missed opportunity--I believe--for you to have done this much work on something Rebol-like and yet not get into the "fun" of those designs. (If I seem puzzled about what you are doing, it is because I know how close you are to those things...but only scratching the surface.)

I have obviously seen the overlap and it's definitely interesting. (And I'm still baffled by how it initially came to happen... I mean Arturo initially looked more like a Tcl variant... until I started looking for ways to get rid of the parentheses 😲) And of course, I would be open to see how things we have both touched have been dealt with (or not)... I'm just not really willing to be influenced in a way that would change what the initial idea about Arturo was. Don't take me wrong: I always appreciate external/fresh ideas... Just not in a way that would alter its essence. Which is why I kind-of-half-avoid the languages of the same big "family". In any case: if you want to chip in any idea/comment/suggestion/food-for-thought, or anything else for that matter, feel free to - you're welcome, of course! 😉 

> Not my typical medium, but I can look at it...

Don't worry! I swear I hadn't even heard of the thing before Arturo... 🤣 It was more out of "necessity" (= how to actually be able to promote the language while engaging with the people that have questions/suggestion/whatever about it) and seeing what other communities do. I'm not a guru by any stretch of the imagination (I still consider myself quite old-school in all of that), but... it's not working for us too bad either...

Have a great day! 😉
